sw/device/lib/testing/test_framework/check.h

```diff
@@ -209,4 +210,27 @@
     }                                                                \
   } while (false)
 
+/**
+ * Checks that the `status_t` represents a non-error value.
+ *
+ * Prints a human-readable error message if the status represents an error.
+ *
+ * @param expr An expression which evaluates to a `status_t`.
+ */
+#define CHECK_STATUS_OK(expr, ...)                                            \
+  do {                                                                        \
+    status_t status_ = expr;                                                  \
+    if (!status_ok(status_)) {                                                \
+      char mod_[3];                                                           \
+      int32_t arg_;                                                           \
+      const char *code_;                                                      \
+      status_extract(status_, &code_, &arg_, mod_);                           \
+      LOG_ERROR(                                                              \
+          "CHECK-STATUS-fail: %s error - generated by module %c%c%c at line " \
+          "%d",                                                               \
+          code_, mod_[0], mod_[1], mod_[2], arg_);                            \
```

Comment:
There is a custom `printf` specifier in our implementation for status codes.  I think you can change this to:
```
LOG_ERROR("CHECK-STATUS-fail: %r", status_);
```

Reply:
As I understand it, the crypto library is free to use `status_t` internally to generate and report errors.  At the API layer, only `crypto_status_t` (from datatypes.h) is returned to callers.

Because we want to avoid calls to the `status_create` function, the internal implementations should not use the normal status creating macros in `//sw/device/lib/base/status.h`, but rather should use `OT_CRYTPO_...` macros.

This all sounds ok to me, but would you object to simply returning `status_t` at the API boundary instead of `crypto_status_t` (or make `crypto_status_t` bit-compatible with `status_t`?)  I'm not seeing what the "crypto" variant gives us over the plain `status_t`?
1. `crypto_status_t` values are explicitly hardened (however, they are 16-bit values, so they don't have the advantage of being loadable as immediate values, which are limited to 12 bits in riscv).
2. `status_t` values include the status code, line number and module-id that generated them.  While not explicitly hardened values, there are lots of bits set over the 32-bit word.
3. I'm not certain if we'd consider the module-id and line number in the status code as revealing too much.  However, since the library is open-source, we're already revealing everything.
4. The module-id and line-number fields of the `status_t` can't be considered stable from version-to-version of the library, but they were never meant to be.  Checking for a specific error in a `status_t` always means extracting it with the `status_code` helper.

If we were to reduce `crypto_status_t` to 12-bit values, then I'd find (1) and (4) to be the most compelling considerations in favor a separate `crypto_status_t`.

Reply:
Hmm, good points. I'll bring it up at the next API meeting to either use `status_t` or at least change to 12-bit values. When `status_t` was first introduced to OT I remember some concern about (3) -- you could certainly construct a cryptographic implementation where the line number is revealing (for instance, a padding oracle, combined with random delays that interfere with the timing properties so the attacker wouldn't know where the error came from without the line number). But I can't think of anything that's not a bit contrived, so maybe a little extra awareness about error leakage in crypto implementations is enough and worth it to make things a bit simpler at the interface.

However, changes to the API are necessarily slow because they have to be agreed upon, go into the API doc, and then get into the next round of manual updates (we're working on a faster process). so I would prefer not to block this PR in the meantime -- if we do end up changing to `status_t` at the top level, it will be pretty straightforward to adjust just the top-level wrappers; most of the code will be below the surface and using `status_t` anyway.

Reply:
One more thing to consider in constructing a `crypto_status_t` (or rather, an elaboration on a previously mentioned idea): you can make `crypto_status_t` compatible with `status_t` so that the crypto library API boundary is `crypto_status_t`, but the conversion into `status_t` is nearly zero cost.  There are two variations on "compatible":

1. Actually bit-for-bit compatible with `status_t` so that the conversion is just a 1-word memcpy (which the compiler will optimize away anyway).
2. Bit-compatibility of the `code` value (ie: `absl_status_t`) so that the conversion is a single bitfield extraction operation.

We use (2) to make`dif_result_t` convertible into `status_t` by way of the `INTO_STATUS` macro in `sw/device/lib/base/status.h`.  This allows test code to use the `TRY` macro with dif functions (really, we should eliminate `dif_result_t` in favor of `status_t`, but that was more work that I wanted to take on.  Similarly, we should convert `rom_error_t` into `status_t`).

If you wanted to go with high hamming-weigh 12-bit values for crypto_status_t, as long as you engineered those values to contain the 5-bit `absl_status_t` code, a conversion to `status_t` would be trivial:

```
typedef enum crypto_status {
  kCryptoStatusOK = (0x46 << 5) | kOk,
  kCryptoStatusBadArgs = (0x69 << 5) | kInvalidArgument,
  kCryptoStatusInternalError = (0x5e << 5) | kInternal,
  //etc...
} crypto_status_t;

// And then, we'd an an else-if stanza to `INTO_STATUS` like this:
} else if (__builtin_types_compatible_p(typeof(ex_), crypto_status_t)) {
    absl_status_t code = (absl_status_t)(ex_ & 0x1F);
    status_ = status_create(code, kStatusModuleId, /* etc */);
}
```

My personal preference is to use `status_t` itself, or to make `crypto_status_t` bit-for-bit compatible, but there still are solutions with nearly zero-cost conversions.  All of this concern over error values is to simplify how our code interacts with error conditions (and specifically to simplify test code).